yw_website/yw_website/apps/website/utils.py

```diff
@@ -89,3 +90,24 @@ def get_direct_descendants(program_block_id, program_blocks):
         new_child.direct_descendants = get_direct_descendants(new_child.id, program_blocks)
         descendants.append(new_child)
     return descendants
+
+def get_ports(parents):
+    for block in parents:
+        for port in Port.objects.filter(on_program_block_id=block.id):
+            new_port = Ports()
+            new_port.name = port.name
+            new_port.id = port.id
+            new_port.is_inport = port.is_inport
+            new_port.is_outport = port.is_outport
+            new_port.data_id = port.data_id
+            new_port.on_program_block_id = port.on_program_block_id
+            new_port.run_id = port.run_id
+            if new_port.is_inport and not new_port.is_outport:
+                block.in_ports.append(new_port)
+            elif new_port.is_outport and not new_port.is_inport:
+                block.out_ports.append(new_port)
+            else: 
+                # both in port and out port
+                # TODO: error handle gracefully if we have a port that is both in and out
```

Comment:
good, we don't wan to raise an error right now (although if we had more time we would want that)